File: refactor.js
const fs = require('fs');
const path = require('path');

const folder = __dirname;
const sidebarHTML = `
    <!-- Sidebar Navigation -->
    <aside class="sidebar">
        <div class="sidebar-header">
            <h2>Architecture Hub</h2>
            <span class="course-code">CSC 2111</span>
        </div>
        
        <nav class="nav-menu">
            <div class="nav-group-title">Overview</div>
            <ul>
                <li class="nav-item"><a href="index.html" class="nav-link"><span class="nav-icon">🏠</span> Dashboard</a></li>
                <li class="nav-item"><a href="terminology.html" class="nav-link"><span class="nav-icon">📖</span> Terminology</a></li>
            </ul>

            <div class="nav-group-title">Topics</div>
            <ul>
                <li class="nav-item"><a href="topic1.html" class="nav-link"><span class="nav-icon">01</span> Structure & Function</a></li>
                <li class="nav-item"><a href="topic2.html" class="nav-link"><span class="nav-icon">02</span> Evolution</a></li>
            </ul>

            <div class="nav-group-title">Assessments</div>
            <ul>
                <li class="nav-item"><a href="quiz_topic1.html" class="nav-link"><span class="nav-icon">🧠</span> Topic 1 Quiz</a></li>
                <li class="nav-item"><a href="quiz_topic2.html" class="nav-link"><span class="nav-icon">🧠</span> Topic 2 Quiz</a></li>
            </ul>
        </nav>
    </aside>
`;

function processHtmlFile(filename, newName, processor) {
    const filePath = path.join(folder, filename);
    if (!fs.existsSync(filePath)) return;
    
    let content = fs.readFileSync(filePath, 'utf-8');
    
    // add link to styles.css
    content = content.replace('</head>', '  <link rel="stylesheet" href="styles.css">\n</head>');
    
    const bodyStart = content.indexOf('<body>') !== -1 ? content.indexOf('<body>') + 6 : content.match(/<body[^>]*>/).index + content.match(/<body[^>]*>/)[0].length;
    const bodyEnd = content.lastIndexOf('</body>');
    
    const pre = content.substring(0, bodyStart);
    const body = content.substring(bodyStart, bodyEnd);
    const post = content.substring(bodyEnd);
    
    processor(pre, body, post, newName);
}

// Topic 1
processHtmlFile('topic1_computer_structure (1).html', 'topic1.html', (pre, body, post, newName) => {
    // Extract Challenge Quiz
    const quizMatch = body.match(/(<!--\s*CHALLENGE QUIZ\s*-->[\s\S]*?)<!--\s*NAV\s*-->/i) || body.match(/(<div class="challenge-section">[\s\S]*?<\/div>\s*)<!--\s*NAV\s*-->/i);
    
    if (quizMatch) {
        const challengeBlock = quizMatch[1];
        let newBody = body.replace(challengeBlock, `
        <div style="margin: 3rem 0; text-align: center;">
            <a href="quiz_topic1.html" style="background: var(--accent); color: #000; padding: 1rem 2rem; border-radius: 50px; font-weight: 700; text-decoration: none; font-family: 'Syne', sans-serif; display: inline-block;">Take Topic 1 Quiz →</a>
        </div>
        `);
        
        newBody = newBody.replace(/alert\('Great work!.*?Next Topic →.*?<\/button>/s, "window.location.href='topic2.html'\">Next Topic →</button>");
        
        const topic1Html = `${pre}\n${sidebarHTML}\n<main class="main-content" style="background: var(--bg); color: var(--text); padding-top: 1rem;">\n${newBody}\n</main>\n${post}`;
        fs.writeFileSync(path.join(folder, newName), topic1Html);
        
        const quizBody = `
        <div class="container" style="padding-top: 1rem;">
            <header>
                <div class="course-tag" style="display:inline-block; font-family:'Space Mono'; color:var(--accent); border:1px solid var(--accent); padding:4px 14px; border-radius:20px; font-size:0.75rem; letter-spacing:2px; margin-bottom:1rem; background:rgba(0,245,196,0.1);">CSC 2111 · ASSESSMENT</div>
                <h1>Topic 1 <br><span>Knowledge Check</span></h1>
                <p style="font-family:'Space Mono'; color:var(--muted); font-size:1rem; margin-top:0.5rem;">// test your understanding</p>
            </header>
            ${challengeBlock}
        </div>
        `;
        const quizHtml = `${pre}\n${sidebarHTML}\n<main class="main-content" style="background: var(--bg); color: var(--text); padding-top: 1rem;">\n${quizBody}\n</main>\n${post}`;
        fs.writeFileSync(path.join(folder, 'quiz_topic1.html'), quizHtml);
    }
});

// Topic 2
processHtmlFile('topic2_evolution_performance.html', 'topic2.html', (pre, body, post, newName) => {
    const quizMatch = body.match(/(<!--\s*QUIZ\s*-->[\s\S]*?<\/div>)\s*<hr>/) || body.match(/(<div class="quiz-section">[\s\S]*?<\/div>\s*<\/div>\s*)<div class="score-bar">/);
    
    if (quizMatch) {
        const challengeBlock = quizMatch[1];
        let newBody = body.replace(challengeBlock, `
        <div style="margin: 3rem 0; text-align: center;">
            <a href="quiz_topic2.html" style="background: var(--accent2); color: #fff; padding: 1rem 2rem; border-radius: 8px; font-weight: 700; text-decoration: none; font-family: 'Outfit', sans-serif; display: inline-block; box-shadow: 3px 3px 0 #1a252f; transition: all 0.2s;">Take Topic 2 Quiz →</a>
        </div>
        `);
        
        newBody = newBody.replace(/alert\('Nice work!.*?Topic 3 →.*?<\/button>/s, "window.location.href='index.html'\">Complete Module →</button>");
        
        const topic2Html = `${pre}\n${sidebarHTML}\n<main class="main-content" style="background: var(--bg); color: var(--text); padding-top: 1rem;">\n${newBody}\n</main>\n${post}`;
        fs.writeFileSync(path.join(folder, newName), topic2Html);
        
        const quizBody = `
        <div class="container" style="padding-top: 1rem;">
            <header>
                <div class="course-tag" style="display:inline-block; font-family:'IBM Plex Mono'; color:var(--accent); font-size:0.75rem; letter-spacing:3px; text-transform:uppercase; margin-bottom:0.8rem;">CSC 2111 · ASSESSMENT</div>
                <h1>Topic 2 <span>Knowledge Check</span></h1>
                <p style="font-family:'IBM Plex Mono'; color:var(--muted); font-size:0.8rem; margin-top:0.5rem;">// verify your understanding of Moore's law and CPI</p>
            </header>
            ${challengeBlock}
            <div class="score-bar" style="display:flex; justify-content:space-between; align-items:center; background:var(--surface); border-radius:10px; padding:1rem 1.5rem; margin:2rem 0; border:1px solid var(--border);">
                <span style="font-family:'IBM Plex Mono'; font-size:0.85rem; color:var(--muted);">Assessment Complete</span>
                <strong id="scoreDisplay" style="color:var(--accent2); font-size:1.1rem;">Score: 0 / 4</strong>
                <button class="next-btn" style="background:var(--accent2); color:#fff; font-family:'Outfit'; font-weight:700; font-size:1rem; padding:0.9rem 2.2rem; border:none; border-radius:8px; cursor:pointer; box-shadow:3px 3px 0 #1a252f;" onclick="window.location.href='index.html'">Finish</button>
            </div>
        </div>
        `;
        const quizHtml = `${pre}\n${sidebarHTML}\n<main class="main-content" style="background: var(--bg); color: var(--text); padding-top: 1rem;">\n${quizBody}\n</main>\n${post}`;
        fs.writeFileSync(path.join(folder, 'quiz_topic2.html'), quizHtml);
    }
});

// Terminology
processHtmlFile('csc2111_terminology.html', 'terminology.html', (pre, body, post, newName) => {
    // just wrap
    const termHtml = `${pre}\n${sidebarHTML}\n<main class="main-content" style="background: var(--bg); color: var(--text); padding-top: 1rem;">\n${body}\n</main>\n${post}`;
    fs.writeFileSync(path.join(folder, newName), termHtml);
});

console.log("Files processed successfully");

File: refactor.py
import re
import os

folder = r"c:\Users\nicholas\Desktop\School\Computer Architecture\Website"
sidebar_html = """
    <!-- Sidebar Navigation -->
    <aside class="sidebar">
        <div class="sidebar-header">
            <h2>Architecture Hub</h2>
            <span class="course-code">CSC 2111</span>
        </div>
        
        <nav class="nav-menu">
            <div class="nav-group-title">Overview</div>
            <ul>
                <li class="nav-item"><a href="index.html" class="nav-link"><span class="nav-icon">🏠</span> Dashboard</a></li>
                <li class="nav-item"><a href="terminology.html" class="nav-link"><span class="nav-icon">📖</span> Terminology</a></li>
            </ul>

            <div class="nav-group-title">Topics</div>
            <ul>
                <li class="nav-item"><a href="topic1.html" class="nav-link"><span class="nav-icon">01</span> Structure & Function</a></li>
                <li class="nav-item"><a href="topic2.html" class="nav-link"><span class="nav-icon">02</span> Evolution</a></li>
            </ul>

            <div class="nav-group-title">Assessments</div>
            <ul>
                <li class="nav-item"><a href="quiz_topic1.html" class="nav-link"><span class="nav-icon">🧠</span> Topic 1 Quiz</a></li>
                <li class="nav-item"><a href="quiz_topic2.html" class="nav-link"><span class="nav-icon">🧠</span> Topic 2 Quiz</a></li>
            </ul>
        </nav>
    </aside>
"""

def split_html(file_path):
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()
    
    # Add link to styles.css in head
    content = content.replace("</head>", '  <link rel="stylesheet" href="styles.css">\n</head>')
    
    # Wrap body in main-content and add sidebar
    # Find <body> tag
    body_start = content.find("<body>")
    if body_start == -1:
        # try with attributes
        body_str = re.search(r'<body[^>]*>', content).group()
        body_start = content.find(body_str)
        body_end_tag = body_start + len(body_str)
    else:
        body_end_tag = body_start + 6
        
    body_end = content.rfind("</body>")
    
    pre_body = content[:body_end_tag]
    body_content = content[body_end_tag:body_end]
    post_body = content[body_end:]
    
    return pre_body, body_content, post_body

def process_topic1():
    file_path = os.path.join(folder, "topic1_computer_structure (1).html")
    pre, body, post = split_html(file_path)
    
    # Extract challenge section
    # the section starts with <div class="challenge-section">
    # and ends after its matching </div>
    # Using regex to find the block (it has specific content)
    match = re.search(r'(<div class="challenge-section">.*?</div>\s*<!-- NAV -->)', body, re.DOTALL)
    if not match:
        match = re.search(r'(<!-- CHALLENGE QUIZ -->.*?</div>\s*<!-- NAV -->)', body, re.DOTALL)
    
    if match:
        challenge_block = match.group(0)
        # remove from main body
        body = body.replace(challenge_block, """
        <div style="margin: 3rem 0; text-align: center;">
            <a href="quiz_topic1.html" style="background: var(--accent); color: #000; padding: 1rem 2rem; border-radius: 50px; font-weight: 700; text-decoration: none; font-family: 'Syne', sans-serif; display: inline-block;">Take Topic 1 Quiz →</a>
        </div>
        <!-- NAV -->
        """)
        
        # update NAV in body
        body = body.replace("alert('Great work! 🎉 Ask me to continue to Topic 2: Computer Evolution & Performance!')", "window.location.href='topic2.html'")
        
        # Create topic1.html
        topic1_html = pre + "\n" + sidebar_html + '\n<main class="main-content" style="background: var(--bg); color: var(--text);">\n' + body + '\n</main>\n' + post
        with open(os.path.join(folder, "topic1.html"), "w", encoding="utf-8") as f:
            f.write(topic1_html)
            
        # Create quiz_topic1.html
        # Extract script that manages quiz
        scripts_match = re.search(r'(<script>.*?</script>)', post, re.DOTALL)
        script_content = scripts_match.group(1) if scripts_match else ""
        
        quiz_body = f"""
        <div class="container">
            <header>
                <div class="course-tag">CSC 2111 · ASSESSMENT</div>
                <h1>Topic 1 <span>Knowledge Check</span></h1>
                <p>// test your understanding</p>
            </header>
            {challenge_block}
        </div>
        """
        quiz_html = pre + "\n" + sidebar_html + '\n<main class="main-content" style="background: var(--bg); color: var(--text);">\n' + quiz_body + '\n</main>\n' + post
        with open(os.path.join(folder, "quiz_topic1.html"), "w", encoding="utf-8") as f:
            f.write(quiz_html)

def process_topic2():
    file_path = os.path.join(folder, "topic2_evolution_performance.html")
    pre, body, post = split_html(file_path)
    
    match = re.search(r'(<!-- QUIZ -->.*?</div>\s*</div>\s*<div class="score-bar">.*?</button>\s*</div>)', body, re.DOTALL)
    
    if match:
        challenge_block = match.group(0)
        body = body.replace(challenge_block, """
        <div style="margin: 3rem 0; text-align: center;">
            <a href="quiz_topic2.html" style="background: var(--accent2); color: #fff; padding: 1rem 2rem; border-radius: 8px; font-weight: 700; text-decoration: none; font-family: 'Outfit', sans-serif; display: inline-block;">Take Topic 2 Quiz →</a>
        </div>
        """)
        
        # Create topic2.html
        topic2_html = pre + "\n" + sidebar_html + '\n<main class="main-content" style="background: var(--bg); color: var(--text);">\n' + body + '\n</main>\n' + post
        with open(os.path.join(folder, "topic2.html"), "w", encoding="utf-8") as f:
            f.write(topic2_html)
            
        # Create quiz_topic2.html
        quiz_body = f"""
        <div class="container">
            <header>
                <span class="course-tag">CSC 2111 · ASSESSMENT</span>
                <h1>Topic 2 <span>Knowledge Check</span></h1>
                <p>// check your understanding</p>
            </header>
            {challenge_block.replace("alert('Nice work! 🎉\\n\\nNext up: Topic 3 — Computer Function & Interconnection.\\nAsk me to continue when ready!')", "window.location.href='index.html'")}
        </div>
        """
        quiz_html = pre + "\n" + sidebar_html + '\n<main class="main-content" style="background: var(--bg); color: var(--text);">\n' + quiz_body + '\n</main>\n' + post
        with open(os.path.join(folder, "quiz_topic2.html"), "w", encoding="utf-8") as f:
            f.write(quiz_html)


def process_terminology():
    file_path = os.path.join(folder, "csc2111_terminology.html")
    pre, body, post = split_html(file_path)
    
    term_html = pre + "\n" + sidebar_html + '\n<main class="main-content" style="background: var(--bg); color: var(--text);">\n' + body + '\n</main>\n' + post
    with open(os.path.join(folder, "terminology.html"), "w", encoding="utf-8") as f:
        f.write(term_html)

process_topic1()
print("Processed topic 1")
process_topic2()
print("Processed topic 2")
process_terminology()
print("Processed terminology")
